data-prepper-plugins/user-agent-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/useragent/UserAgentProcessor.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.processor.useragent;
+
+import org.opensearch.dataprepper.metrics.PluginMetrics;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPlugin;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPluginConstructor;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.processor.AbstractProcessor;
+import org.opensearch.dataprepper.model.processor.Processor;
+import org.opensearch.dataprepper.model.record.Record;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import ua_parser.CachingParser;
+import ua_parser.Client;
+import ua_parser.Parser;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.opensearch.dataprepper.logging.DataPrepperMarkers.EVENT;
+
+@DataPrepperPlugin(name = "user_agent", pluginType = Processor.class, pluginConfigurationType = UserAgentProcessorConfig.class)
+public class UserAgentProcessor extends AbstractProcessor<Record<Event>, Record<Event>> {
+    private static final Logger LOG = LoggerFactory.getLogger(UserAgentProcessor.class);
+    private final UserAgentProcessorConfig config;
+    private final Parser userAgentParser;
+
+    @DataPrepperPluginConstructor
+    public UserAgentProcessor(final PluginMetrics pluginMetrics, final UserAgentProcessorConfig config) {
+        super(pluginMetrics);
+        this.config = config;
+        this.userAgentParser = new CachingParser();
+    }
+
+    @Override
+    public Collection<Record<Event>> doExecute(final Collection<Record<Event>> records) {
+        for (final Record<Event> record : records) {
+            final Event event = record.getData();
+
+            final String userAgentStr = event.get(config.getSource(), String.class);
+            if (Objects.isNull(userAgentStr)) {
+                continue;
+            }
+
+            try {
+                final Client clientInfo = this.userAgentParser.parse(userAgentStr);
+
+                final Map<String, Object> parsedUserAgent = getParsedUserAgent(clientInfo);
+                if (!config.getExcludeOriginal()) {
+                    parsedUserAgent.put("original", userAgentStr);
+                }
+                event.put(config.getTarget(), parsedUserAgent);
+            } catch (Exception e) {
+                LOG.error(EVENT, "An exception occurred when parsing user agent data from event [{}] with source key [{}]",
```

Comment:
We should probably add an option to tag this. See #2682 for a similar feature which was just added to Grok.

You don't have to include this in this PR. But, if not, please create a GitHub issue to track it. Thanks!

Reply:
Added `tags_on_parse_failure` option.